test/AutoDiff/leakchecking.swift

```diff
@@ -215,6 +215,156 @@ LeakCheckingTests.testWithLeakChecking("ClassMethods") {
+
+struct TF_508_Struct<Scalar : AdditiveArithmetic>
+  : TF_508_Proto, AdditiveArithmetic {}
+extension TF_508_Struct : Differentiable where Scalar : Differentiable {
+  typealias TangentVector = TF_508_Struct
+}
+
+// Test leaks regarding `SILGenFunction::getOrCreateAutoDiffLinearMapThunk`.
+LeakCheckingTests.testWithLeakChecking("LinearMapSILGenThunks") {
+  func testLinearMapSILGenThunks() {
+    let x = TF_508_Struct<Float>()
+    // Test conformance requirement with dependent member type.
+    _ = pullback(at: x, in: { (x: TF_508_Struct<Float>) -> TF_508_Struct<Float> in
+      return x + x
+    })
+    // Test same-type requirement with dependent member type.
+    _ = pullback(at: x, in: { (x: TF_508_Struct<Float>) -> TF_508_Struct<Float> in
+      return x.subtract(x)
+    })
+  }
+  testLinearMapSILGenThunks()
+}
+
+LeakCheckingTests.testWithLeakChecking("ParameterConventionMismatchLeakChecking") {
+  struct Nontrivial : Differentiable {
+    var base: Tracked<Float>
+
+    // Test initializer and static VJP function.
+    // Initializers have owned parameters but functions have shared parameters.
+    @differentiable(vjp: vjpInit)
+    init(_ base: Tracked<Float>) {
+      self.base = base
+    }
+    static func vjpInit(_ base: Tracked<Float>)
+      -> (Nontrivial, (TangentVector) -> Tracked<Float>) {
+      return (Nontrivial(base), { v in v.base })
+    }
+
+    @differentiable(vjp: vjpOwnedParameterMismatch)
+    func ownedParameter(_ x: __owned Tracked<Float>) -> Tracked<Float> {
+      return x
+    }
+    func vjpOwnedParameterMismatch(_ x: __shared Tracked<Float>)
+      -> (Tracked<Float>, (Tracked<Float>) -> (TangentVector, Tracked<Float>)) {
+      return (ownedParameter(x), { v in (.zero, v) })
+    }
+
+    @differentiable(vjp: vjpSharedParameterMismatch)
+    func sharedParameter(_ x: __shared Tracked<Float>) -> Tracked<Float> {
+      return x
+    }
+    func vjpSharedParameterMismatch(_ x: __owned Tracked<Float>)
+      -> (Tracked<Float>, (Tracked<Float>) -> (TangentVector, Tracked<Float>)) {
+      return (sharedParameter(x), { v in (.zero, v) })
+    }
+
+    @differentiable(vjp: vjpOwnedParameterGenericMismatch)
+    func ownedParameterGeneric<T : Differentiable>(_ x: __owned T) -> T {
+      return x
+    }
+    func vjpOwnedParameterGenericMismatch<T : Differentiable>(_ x: __shared T)
+      -> (T, (T.TangentVector) -> (TangentVector, T.TangentVector)) {
+      return (ownedParameterGeneric(x), { v in (.zero, v) })
+    }
+
+    @differentiable(vjp: vjpSharedParameterGenericMismatch)
+    func sharedParameterGeneric<T : Differentiable>(_ x: __shared T) -> T {
+      return x
+    }
+    func vjpSharedParameterGenericMismatch<T : Differentiable>(_ x: __owned T)
+      -> (T, (T.TangentVector) -> (TangentVector, T.TangentVector)) {
+      return (sharedParameterGeneric(x), { v in (.zero, v) })
+    }
+
+    @differentiable(vjp: vjpConsuming)
+    __consuming func consuming(_ x: Tracked<Float>) -> Tracked<Float> {
```

Comment:
Also add tests where the original is non-`__consuming` but the JVP and VJP are.

Reply:
Done in https://github.com/apple/swift/pull/26380/commits/bba4b634e7ad23a0ce165746cfd074a66ad7ceb1.